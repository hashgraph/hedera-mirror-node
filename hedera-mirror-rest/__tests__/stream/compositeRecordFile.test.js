/*-
 * ‌
 * Hedera Mirror Node
 * ​
 * Copyright (C) 2019 - 2021 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

'use strict';

const {CompositeRecordFile} = require('../../stream');
const {testRecordFiles, commonRecordFileTests, copyRecordFileAndSetVersion} = require('./testUtils');

describe('unsupported record file version', () => {
  const bufferV2 = testRecordFiles.v2[0].buffer;
  // unsupported version numbers 3, 4, and 6 (future version)
  const testSpecs = [...[3, 4, 6].map((version) => [version, copyRecordFileAndSetVersion(bufferV2, version)])];

  testSpecs.forEach((testSpec) => {
    const [version, buffer] = testSpec;
    test(`create object from version ${version}`, () => {
      expect(() => new CompactRecordFile(buffer)).toThrowErrorMatchingSnapshot();
    });
  });
});

describe('canCompact', () => {
  const bufferV2 = testRecordFiles.v2[0].buffer;
  const testSpecs = [
    [1, false],
    [2, false],
    [3, false],
    [4, false],
    [5, true],
    [6, false],
  ].map(([version, expected]) => {
    return {
      version,
      buffer: copyRecordFileAndSetVersion(bufferV2, version),
      expected,
    };
  });

  testSpecs.forEach((testSpec) => {
    const {version, buffer, expected} = testSpec;
    test(`version ${version} - ${expected ? 'can compact' : 'cannot compact'}`, () => {
      expect(CompositeRecordFile.canCompact(buffer)).toEqual(expected);
    });
  });
});

describe('from record file buffer or compact object', () => {
  commonRecordFileTests(2, CompositeRecordFile, false);
  commonRecordFileTests(5, CompositeRecordFile, true);
});
