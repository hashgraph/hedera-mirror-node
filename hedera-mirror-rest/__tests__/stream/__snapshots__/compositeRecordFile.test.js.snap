// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`from record file buffer or compact object check individual field from v2 buffer - toCompactObject - ["0.0.365299-1615141267-266970662"] 1`] = `"Unsupported operation"`;

exports[`from record file buffer or compact object check individual field from v5 buffer - toCompactObject - ["0.0.1-123-123456789"] 1`] = `"Transaction 0.0.1-123-123456789 not found in the successful transactions map"`;

exports[`from record file buffer or compact object check individual field from v5 compact object - toCompactObject - ["0.0.1-123-123456789"] 1`] = `"Transaction 0.0.1-123-123456789 not found in the successful transactions map"`;

exports[`from record file buffer or compact object check individual field from v6 buffer - toCompactObject - ["0.0.1-123-123456789"] 1`] = `"Transaction 0.0.1-123-123456789 not found in the successful transactions map"`;

exports[`from record file buffer or compact object check individual field from v6 compact object - toCompactObject - ["0.0.1-123-123456789"] 1`] = `"Transaction 0.0.1-123-123456789 not found in the successful transactions map"`;

exports[`from record file buffer or compact object end running hash mismatch 1`] = `"End object running hash mismatch"`;

exports[`from record file buffer or compact object end running hash mismatch 2`] = `"End object running hash mismatch"`;

exports[`from record file buffer or compact object from non-Buffer obj 1`] = `"Cannot read properties of undefined (reading 'readInt32BE')"`;

exports[`from record file buffer or compact object truncated v2 buffer 1`] = `"Error reading byte array, expect 221-byte data got 220-byte"`;

exports[`from record file buffer or compact object truncated v5 buffer 1`] = `"Error reading byte array, expect 48-byte data got 47-byte"`;

exports[`from record file buffer or compact object truncated v6 buffer 1`] = `"index out of range: 6658 + 1 > 6658"`;

exports[`from record file buffer or compact object v2 buffer with extra data 1`] = `"Unsupported marker 0, expect 2"`;

exports[`from record file buffer or compact object v5 buffer with extra data 1`] = `"Extra data discovered in record file"`;

exports[`from record file buffer or compact object v6 buffer with extra data 1`] = `"index out of range: 6660 + 1 > 6660"`;

exports[`unsupported record file version create object from version 3 1`] = `"Unsupported record file"`;

exports[`unsupported record file version create object from version 4 1`] = `"Unsupported record file"`;

exports[`unsupported record file version create object from version 7 1`] = `"Unsupported record file"`;
